Grid Visualizer - improve styles

talldan opened the pull request:
## What?
See #64162

Brings the styles of the grid visualizer closer to those shown in the video on the issue.

There are a few bits this doesn't do from the video:
- The grid isn't revealed on hover, I feel this is better to do in a separate PR
- Dragging to select multiple grid cells is also out of scope for this PR
- In auto mode, the video shows the next available cell being interactive like the manual mode cells. This should be possible in a follow up. 

There are a few other subtleties that might worth trying in separate PRs:
- There's quite a lot of visual noise if you have a block with the visualizer squares under it. We could try de-emphasizing the cells under the selected child block within a grid or not showing them at all. The video possibly shows something like this at 1:55, but hard to say for sure

## How?
In order to achieve the visual styling shown in #64162, the grid visualizer needs to be rendered behind the grid block.

This PR attempts this using an inline popover.

It requires quite a bit of code-reshuffling as there are now two separate grid visualizers being rendered:
- `GridPopunder` - renders a visualizer under the selected grid block. This visualizer is entirely visual, it renders some blue rectangles.
- `GridPopover` - renders interactive elements over the selected block. This is responsible for the drop zones and appenders in the grid. It has to be over the selected block to receive mouse events.

## Testing Instructions
1. Add a grid block
2. Switch between auto and manual and notice the cells are blue in manual mode to indicate interactivity
3. In manual mode, try dragging a block into an empty cell and observe the highlight styles.
4. Hover over an empty cell and observe the appender styles

## Screenshots or screencast <!-- if applicable -->
![Screenshot 2024-08-09 at 5 56 41 PM](https://github.com/user-attachments/assets/231f3fa7-2f98-4806-a7dc-050209537e31)

talldan committed:
Add support to root block list in `useBlockRefs`

talldan committed:
Add basic underlay component

talldan committed:
Only render GridVisualizer above GridBlock and no longer on LayoutChild. Use context to provide resizer bounds

talldan committed:
Remove outdated comment

talldan committed:
Move padding to underlay to ensure it refreshes

talldan committed:
Handle block moving

talldan committed:
Add a background for cells

talldan committed:
Use BlockPopoverCover with inline prop

talldan committed:
Allow margin reset on inline popover via inline styles

talldan committed:
Override popover default z index

talldan committed:
Render two "grid visualizers", one under the block and one over-the one over has all the interactive elements

talldan committed:
Revert change to root block list

talldan commented:
> Block Movers and Drop Zones are now underneath the grid block so are not receiving mouse events. They probably need to be rendered over the grid block, so it might be worth trying a hybrid approach with some of this PR and some of what's in trunk. 🤔

I've pushed a few more commits that try this, and it works pretty well, it partially solves the issue with the cells inside the editor iframe not having access to component CSS, as now the most interactive parts of the grid visualizer are outside the canvas, and only the cells underneath the grid need some basic styles which might be as simple as inlines styles.

noisysocks commented:
This is cool. I'll take it for a spin.

> * It might be worth experimenting with shadow dom to prevent styles leaking into the canvas, and then perhaps it'll possible to add stylesheets within that container.

Do you think it's still worth doing this? Or are the styles minimal enough now that it doesn't matter?

It's a problem we run into semi frequently. Might be fun to build an `<Isolated>` component that puts its children into shadow DOM or something along those lines.

noisysocks commented:
This is really cool—you're a genius.

Can we hide the drop zone / inserter on occupied cells when you're not dragging so that you don't see stripes on top of blocks? 

<img width="501" alt="Screenshot 2024-08-09 at 12 37 18" src="https://github.com/user-attachments/assets/02960ed3-9133-4ba5-b04d-892de08cd7da">

talldan committed:
Update grid styles

talldan committed:
Update appender styles

talldan committed:
Hide drop zone until active

talldan commented:
> Can we hide the drop zone / inserter on occupied cells when you're not dragging so that you don't see stripes on top of blocks?

I was wondering why some cells had a slightly different color. 😄 

I've pushed a commit that does that, and also a bunch more styling updates to make it look more like the video in https://github.com/WordPress/gutenberg/issues/64162#issue-2441644988. I'm mostly doing this by sight, but I think it's pretty close.

I'll work a bit more on the highlighted cell styles. Then I probably need to tidy up some before it's ready for review.

I'm surprised how quickly this has come together. That the popover component now properly supports being inline helps a lot.

talldan committed:
Improve highlight style

talldan committed:
Refactor into separate GridPopunder / GridPopover components

talldan committed:
Fix resizer bounds

talldan commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -68,60 +67,112 @@ const GridVisualizerGrid = forwardRef(
+ * @param {Object}  props.gridInfo
+ * @param {boolean} props.isManualGrid
+ */
+function GridPopunder( { gridClientId, gridInfo, isManualGrid } ) {
```
Not sure if `GridPopunder` is too twee in terms of naming. 😄

talldan committed:
Try for a smaller diff

talldan commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -68,60 +67,59 @@ const GridVisualizerGrid = forwardRef(
-			}
-			function onGlobalDragEnd() {
-				setIsDroppingAllowed( false );
-			}
-			document.addEventListener( 'drag', onGlobalDrag );
-			document.addEventListener( 'dragend', onGlobalDragEnd );
-			return () => {
-				document.removeEventListener( 'drag', onGlobalDrag );
-				document.removeEventListener( 'dragend', onGlobalDragEnd );
-			};
-		}, [] );
-
 		return (
-			<BlockPopoverCover
-				className={ clsx( 'block-editor-grid-visualizer', {
-					'is-dropping-allowed': isDroppingAllowed,
-				} ) }
-				clientId={ gridClientId }
-				__unstablePopoverSlot="__unstable-block-tools-after"
-			>
-				<div
-					ref={ ref }
-					className="block-editor-grid-visualizer__grid"
-					style={ gridInfo.style }
-				>
-					{ isManualGrid ? (
-						<ManualGridVisualizer
-							gridClientId={ gridClientId }
-							gridInfo={ gridInfo }
-						/>
-					) : (
-						Array.from( { length: gridInfo.numItems }, ( _, i ) => (
-							<GridVisualizerCell
-								key={ i }
-								color={ gridInfo.currentColor }
-							/>
-						) )
-					) }
-				</div>
-			</BlockPopoverCover>
```
This has all moved into `GridPopover`

talldan commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -68,60 +67,59 @@ const GridVisualizerGrid = forwardRef(
+ * @param {string} props.gridClientId
+ * @param {Object} props.gridInfo
+ */
+const GridPopover = forwardRef( ( { gridClientId, gridInfo }, ref ) => {
```
`ManualGridVisualizer` is now renamed to `GridPopover`, and it now encompasses all the interactive aspects of the grid visualizer. It renders it's own popover over the grid.

talldan commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -146,56 +144,117 @@ function ManualGridVisualizer( { gridClientId, gridInfo } ) {
-}
+} );
 
-function GridVisualizerCell( { color, children, className } ) {
```
I decided to get rid of `GridVisualizerCell` as it wasn't doing very much, especially in `GridPopover` where it's just a `<div>`.

In `GridPopunder` it's a `<div>` with a some styles, so I think it still doesn't need its own component.

talldan commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -146,56 +144,117 @@ function ManualGridVisualizer( { gridClientId, gridInfo } ) {
+ * @param {Object}  props.gridInfo
+ * @param {boolean} props.isManualGrid
+ */
+function GridPopunder( { gridClientId, gridInfo, isManualGrid } ) {
```
`GridPopunder` (not sure if the name is too twee) now renders the blue cells in the background. It's pretty similar to the auto mode cells in `trunk` in terms of code.

talldan commented on packages/block-editor/src/hooks/layout-child.js:
```diff
@@ -200,11 +196,6 @@ function ChildLayoutControlsPure( { clientId, style, setAttributes } ) {
 
 	return (
 		<>
-			<GridVisualizer
-				clientId={ rootClientId }
-				contentRef={ setResizerBounds }
-				parentLayout={ parentLayout }
-			/>
```
In order to have the `GridPopunder` always render right before the grid block, I couldn't render `GridVisualizer` as it was before as part of the child block (when a child block is selected).

I suppose it is possible to still render `GridPopover` here though, but the downside would be having to calculate `gridInfo` twice. The upside is the `GridResizerBoundsContext` wouldn't be needed. 🤷

noisysocks replied:
I like this approach regardless. Nice to have only one instance of `GridVisualizer`.

talldan commented on packages/components/src/popover/index.tsx:
```diff
@@ -139,6 +139,7 @@ const UnforwardedPopover = (
 		shift = false,
 		inline = false,
 		variant,
+		contentStyle,
```
Probably need to spin the popover changes out into a separate PR. Or look at alternative options for popover styles inside the canvas.

noisysocks replied:
Could it be done in the `style.scss`? I suppose not because it's in the iframe.

talldan replied:
Yep, that's correct, it won't work because of the iframe. Two of the the options are to use inline styles or render a stylesheet. 

Another would be to render the `BlockPopoverCover` inside `<div>` and use that div to reset the margin, but there'd still be the popover z-index to unset. Maybe though the components team would be more open to a `noZIndex` prop or something then though. I'll find out.

noisysocks committed:
Rename GridVisualizerGrid, move occupiedRects

noisysocks commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -68,60 +67,59 @@ const GridVisualizerGrid = forwardRef(
+				/>
+				{ isManualGrid && (
+					<GridPopover
+						ref={ ref }
```
There's a bug here caused by `ref` only being passed/set when `isManualGrid` is true. It means that you're able to resize a grid item beyond the bounds of the grid block when it's in auto mode.

https://github.com/user-attachments/assets/ede25412-04c1-47a4-a3d9-de2c2fa3df37

talldan replied:
Should be fixed in f73ca9f. The annoying aspect is that it means some parts of `GridPopover` always have to be rendered even when there's not much going on, but I think it's necessary. I tried to optimize it by moving the items into a separate component.

To properly finish https://github.com/WordPress/gutenberg/issues/64162, the auto grid will also need a single appender in the next available cell, so something like this will be needed anyway.

noisysocks replied:
> To properly finish https://github.com/WordPress/gutenberg/issues/64162, the auto grid will also need a single appender in the next available cell, so something like this will be needed anyway.

I think we can ignore that tbh. It's better imo to have it consistent across container blocks.

noisysocks commented on packages/block-editor/src/components/grid/style.scss:
```diff
@@ -48,17 +44,26 @@
+		box-shadow: none;
+		background-color: rgba(var(--wp-admin-theme-color--rgb), 0.05);
+		border: 1px dashed;
+		border-radius: 2px;
```
I think we have a `$border-radius` var for this.

talldan replied:
Oh yep, that was lazy. Now updated.

noisysocks commented on packages/block-editor/src/hooks/grid-visualizer.js:
```diff
@@ -4,52 +4,65 @@
 const addGridVisualizerToBlockEdit = createHigherOrderComponent(
 	( BlockEdit ) => ( props ) => {
+		// Use useState() instead of useRef() so that GridItemResizer updates when ref is set.
+		const [ resizerBounds, setResizerBounds ] = useState();
```
Maybe move this `useState` and the `Provider` to a new component so that we're not calling `useState` on every single block regardless of whether it has a grid layout. Probably doesn't make any difference to performance in practice but I think it's a nice habit to be really careful about what goes above the early return in a `BlockEdit` filter.

talldan replied:
Good point, I think it's also much tidier to move it into `GridTools` and have `GridTools` accept `children`. Done in 6139488

noisysocks commented on packages/block-editor/src/components/grid/grid-visualizer.js:
```diff
@@ -146,56 +144,117 @@ function ManualGridVisualizer( { gridClientId, gridInfo } ) {
+	return (
+		<BlockPopoverCover
+			__unstablePopoverSlot="__unstable-block-tools-after"
+			className={ clsx( 'block-editor-grid-visualizer', {
```
The class names in this file have come out of sync with the component names. It used to be that `GridVisualizer` rendered an element with `className = 'block-editor-grid-visualizer'` and so on. Should we update them?

Also, is it bad that both `BlockPopover` and `BlockPopunder` render an element with the same classname?

talldan replied:
Good point. I've pushed a commit that updates these. `GridPopunder` doesn't need classnames it relies on the inline css, so removing those solves the duplication. I changed the `GridPopover` classnames to `block-editor-grid-visualizer-popover` to match the components. 

Also removed another unused classname.

It's not perfect, but I think good enough right now.

noisysocks reviewed (commented):
Love this, it's a big improvement in UI and code quality.

talldan committed:
Update grid css

- Remove unused classnames
- Update existing classnames to use BEM component convention

talldan committed:
Move GridResizerBoundsContext.Provider rendering to `GridTools` and make `GridTools` accept children.

talldan committed:
Use scss variables

talldan committed:
Rename popover prop to `style` instead of `contentStyle`

talldan committed:
Fix: Resizer bounds not working for auto grids

talldan commented:
I've made a separate PR for the popover changes - [Components: Allow style prop on Popover](https://github.com/WordPress/gutenberg/pull/64489#top)

noisysocks reviewed (approved):
Nice work, this is great.

talldan commented:
Unfortunately I've found a couple of bugs that may be difficult to solve.

Rendering the visualizer popover inline before the grid means layout rules can apply to it, like this one:
```css
.is-layout-constrained > :first-child {
    margin-block-start: 0;
}
```

The problem is not so much that the layout rule applies to the visualizer - it can be overridden, more that it no longer applies to the grid block itself like its supposed to. It results in the layout being incorrect. The popover also only renders when the grid is selected, and that causes a layout shift as the css rule jumps from applying to the block to suddenly applying to the popover:

https://github.com/user-attachments/assets/c48c4b75-3032-462e-a73c-228b34d3f6ab

The second issue is that grid block's background can cover the visualizer popover.

Rendering the visualizer popover inside the Grid block is probably the next best option, though there's no easy way to do it using a filter from what I understand 🤔

noisysocks commented:
> Rendering the visualizer popover inside the Grid block is probably the next best option, though there's no easy way to do it using a filter from what I understand 🤔

Can we modify the `children` prop that comes back from `BlockEdit`? Never tried. It's probably a bad idea 😛

noisysocks commented:
Or `createPortal()` into the element returned from `useBlockElement`. I have no shortage of bad ideas 😀

talldan commented:
Or maybe render a slot in the selected grid block. It'd be implemented in the group block, but the slot/fill could be a private API from the block editor package. 🤔

ramonjd commented:
Just wondering if this PR still valid or worth reviving?
